src/jitter_rng.rs

```diff
@@ -0,0 +1,591 @@
+        // following sanity checks verify that we have a high-resolution timer.
+
+        let mut delta_sum = 0;
+        let mut old_delta = 0;
+
+        let mut time_backwards = 0;
+        let mut count_mod = 0;
+        let mut count_stuck = 0;
+
+        // TESTLOOPCOUNT needs some loops to identify edge systems.
+        // 100 is definitely too little.
+        const TESTLOOPCOUNT: u32 = 300;
+        const CLEARCACHE: u32 = 100;
+
+        for i in 0..(CLEARCACHE + TESTLOOPCOUNT) {
+            let time = (self.timer)();
+            self.prev_time = time;
+            // Do some work, invoke core entropy collection logic
+            black_box(self.lfsr_time(time));
+            let time2 = (self.timer)();
+
+            // test whether timer works
+            if time == 0 || time2 == 0 {
+                return Err(new_error(ErrorKind::NoTimer));
+            }
+            let delta = time2.wrapping_sub(time) as i64;
+
+            // Test whether timer is fine grained enough to provide delta even
+            // when called shortly after each other -- this implies that we also
+            // have a high resolution timer
+            if delta == 0 {
+                return Err(new_error(ErrorKind::CoarseTimer));
+            }
+
+            // Up to here we did not modify any variable that will be
+            // evaluated later, but we already performed some work. Thus we
+            // already have had an impact on the caches, branch prediction,
+            // etc. with the goal to clear it to get the worst case
+            // measurements.
+            if i < CLEARCACHE { continue; }
+
+            if self.stuck(delta) { count_stuck += 1; }
+
+            // Test whether we have an increasing timer.
+            if !(time2 > time) { time_backwards += 1; }
+
+            // Count the number of times the counter increases in steps of 100ns
+            // or greater.
+            if (delta % 100) == 0 { count_mod += 1; }
+
+            // Ensure that we have a varying delta timer which is necessary for
+            // the calculation of entropy -- perform this check only after the
+            // first loop is executed as we need to prime the old_delta value
+            delta_sum += (delta - old_delta).abs();
+            old_delta = delta;
+        }
+
+        // We allow the time to run backwards for up to three times.
+        // This can happen if the clock is being adjusted by NTP operations.
+        // If such an operation just happens to interfere with our test, it
+        // should not fail. The value of 3 should cover the NTP case being
+        // performed during our test run.
+        if time_backwards > 3 {
+            return Err(new_error(ErrorKind::NotMonotonic));
+        }
+
+        // Variations of deltas of time must on average be larger than 1 to
+        // ensure the entropy estimation implied with 1 is preserved
+        if delta_sum <= 1 {

```

Comment:
On average? So should the RHS be `TESTLOOPCOUNT`?

Reply:
Yes, good catch.